OrbitCore/FunctionFramepointerValidator.cpp

```diff
@@ -19,113 +19,119 @@ FunctionFramepointerValidator::~FunctionFramepointerValidator() {
-      cs_regs regs_read_first, regs_write_first, regs_read_second,
-          regs_write_second;
-      uint8_t read_count_first, write_count_first, read_count_second,
-          write_count_second;
-      // check for "mov esp, ebp; pop ebp"
-      if (  // first instruction
-          cs_regs_access(handle_, &instructions_[i], regs_read_first,
-                         &read_count_first, regs_write_first,
-                         &write_count_first) == 0 &&
-          read_count_first == 1 && isBasePointer(regs_read_first[0]) &&
-          write_count_first == 1 && isStackPointer(regs_write_first[0]) &&
-          // second instructions
-          cs_regs_access(handle_, &instructions_[i + 1], regs_read_second,
-                         &read_count_second, regs_write_second,
-                         &write_count_second) == 0 &&
-          read_count_second == 1 && isStackPointer(regs_read_second[0]) &&
-          write_count_second == 2 && isStackPointer(regs_write_second[0]) &&
-          isBasePointer(regs_write_second[1])) {
-        found_epilogue = true;
-      }
-    }
-  }
-  return found_epilogue;
-}
-
-bool FunctionFramepointerValidator::validatePrologue() {
+bool FunctionFramepointerValidator::ValidatePrologue() {
   cs_regs regs_read, regs_write;
   uint8_t read_count, write_count;
 
   // check the first instruction: it must be "push ebp"
   if (cs_regs_access(handle_, &instructions_[0], regs_read, &read_count,
                      regs_write, &write_count) == 0) {
     if (instructions_[0].id != X86_INS_PUSH || read_count != 2 ||
-        !isStackPointer(regs_read[0]) || !isBasePointer(regs_read[1])) {
+        !IsStackPointer(regs_read[0]) || !IsBasePointer(regs_read[1])) {
       return false;
     }
   }
 
   // check the second instruction: it must be "mov ebp, esp"
   if (cs_regs_access(handle_, &instructions_[1], regs_read, &read_count,
                      regs_write, &write_count) == 0) {
-    if (!isMovInstruction(instructions_[1]) || read_count != 1 ||
-        !isStackPointer(regs_read[0]) || write_count != 1 ||
-        !isBasePointer(regs_write[0])) {
+    if (!IsMovInstruction(instructions_[1]) || read_count != 1 ||
+        !IsStackPointer(regs_read[0]) || write_count != 1 ||
+        !IsBasePointer(regs_write[0])) {
       return false;
     }
   }
 
   return true;
 }
 
-bool FunctionFramepointerValidator::validateFramePointers() {
-  return validateEpilogue() && validatePrologue();
+// We will actually only check, whether there is any correct epilogue.
+// It might be the case, that there are multiple function returns and that not
+// all are correct. However, we would not expect a compiler to produce this,
+// and for hand written assembly, we accept wrong unwinding results.
+bool FunctionFramepointerValidator::ValidateEpilogue() {
+  bool found_epilogue = false;
+  for (size_t i = 0; i < instructions_count_ - 1; i++) {
+    if (IsMovInstruction(instructions_[i]) &&
```

Comment:
here you can do
if (!IsMov ... ) {
  continue;
}

Reply:
Good point